skimage/restoration/_nl_means_denoising.pyx

```diff
@@ -580,85 +520,90 @@ def _fast_nl_means_denoising_2d(image, int s=7, int d=13, double h=0.1,
     Jacques Froment. Parameter-Free Fast Pixelwise Non-Local Means
     Denoising. Image Processing On Line, 2014, vol. 4, p. 300-326.
     """
+
+    cdef np_floats DISTANCE_CUTOFF = 5.0
     if s % 2 == 0:
         s += 1  # odd value for symmetric patch
-    cdef int offset = s / 2
+
     # Image padding: we need to account for patch size, possible shift,
     # + 1 for the boundary effects in finite differences
-    cdef int pad_size = offset + d + 1
-    cdef IMGDTYPE [:, :, ::1] padded = np.ascontiguousarray(np.pad(image,
-                          ((pad_size, pad_size), (pad_size, pad_size), (0, 0)),
-                          mode='reflect').astype(np.float64))
-    cdef IMGDTYPE [:, :, ::1] result = np.zeros_like(padded)
-    cdef IMGDTYPE [:, ::1] weights = np.zeros_like(padded[..., 0], order='C')
-    cdef IMGDTYPE [:, ::1] integral = np.empty_like(padded[..., 0], order='C')
-    cdef int n_row, n_col, t_row, t_col, row, col, n_channels, channel
-    cdef double weight, distance
-    cdef double alpha
-    cdef double h2 = h * h
-    cdef double s2 = s * s
-    n_row, n_col, n_channels = image.shape
-    cdef double h2s2 = n_channels * h2 * s2
-    n_row += 2 * pad_size
-    n_col += 2 * pad_size
+    cdef Py_ssize_t n_row, n_col, t_row, t_col, row, col, n_channels, channel
+    cdef Py_ssize_t row_start, row_end, row_shift, col_shift
+    cdef Py_ssize_t offset = s / 2
+    cdef Py_ssize_t pad_size = offset + d + 1
+
+    cdef np_floats [:, :, ::1] padded = np.ascontiguousarray(
+        np.pad(image, ((pad_size, pad_size), (pad_size, pad_size), (0, 0)),
+               mode='reflect'))
+    cdef np_floats [:, :, ::1] result = np.zeros_like(padded)
+    cdef np_floats [:, ::1] weights = np.zeros_like(padded[..., 0], order='C')
+    cdef np_floats [:, ::1] integral = np.zeros_like(padded[..., 0], order='C')
```

Comment:
`integral` was previously initialized using `empty_like` rather than `zeros_like` which has slightly less overhead. Is there a reason this now needs to be zeros?

Reply:
On followup, I see that you have removed the resetting `integral[:, :] = 0` below. 

On first glance this seems to be okay, but probably worth double checking. I'm not sure our test cases are super strict. Can you manually verify that the PSNR values and output of [this demo](https://scikit-image.org/docs/dev/auto_examples/filters/plot_nonlocal_means.html#non-local-means-denoising-for-preserving-textures) still look consistent?

Reply:
Yes, that's it, `integral` was set to zero inside the loop instead of being directly initialized to zero.

Reply:
Great. Thanks for checking